src/aapns/connection.py

```diff
@@ -0,0 +1,454 @@
+                # * max_concurrent_streams change
+                # * [maybe] starting a stream
+                # * a stream getting closed (but not half-closed)
+                # * closing / closed change
+        except ConnectionError as e:
+            if not self.outcome:
+                self.outcome = str(e)
+        except Exception:
+            logger.exception("background read task died")
+        finally:
+            self.closing = self.closed = True
+            for channel in self.channels.values():
+                channel.wakeup.set()
+
+    async def background_write(self):
+        try:
+            while not self.closed:
+                data = None
+
+                while not data:
+                    if self.closed:
+                        return
+
+                    if data := self.protocol.data_to_send():
+                        self.write_stream.write(data)
+                        last_stream_id = self.last_stream_id_got
+                        await self.write_stream.drain()
+                        self.last_stream_id_sent = last_stream_id
+                    else:
+                        await self.should_write.wait()
+                        self.should_write.clear()
+
+        except (SSLError, ConnectionError) as e:
+            if not self.outcome:
+                self.outcome = str(e)
+        except Exception:
+            logger.exception("background write task died")
+        finally:
+            self.closing = self.closed = True
+
+
+@dataclass
+class Channel:
+    wakeup: asyncio.Event = field(default_factory=asyncio.Event)
+    events: List[h2.events.Event] = field(default_factory=list)
+    header: Optional[dict] = None
+    body: bytes = b""
+
+
+@dataclass
+class Request:
+    header: tuple
+    body: bytes
+    deadline: float
+
+    def header_with(self, host: str, port: int) -> tuple:
+        """Request header including :authority pseudo header field for target server"""
+        return ((":authority", f"{host}:{port}"),) + self.header
+
+    @classmethod
+    def new(
+        cls,
+        path: str,
+        header: Optional[dict],
+        data: dict,
+        timeout: Optional[float] = None,
+        deadline: Optional[float] = None,
+    ):
+        if timeout is not None and deadline is not None:
+            raise ValueError("Specify timeout or deadline, but not both")
+        elif timeout is not None:
+            deadline = time() + timeout
+        elif deadline is None:
+            deadline = inf
+
+        if not path.startswith("/"):
+            raise ValueError("Absolute URL path is required")
+        pseudo = dict(method="POST", scheme="https", path=path)
+        h = tuple((f":{k}", v) for k, v in pseudo.items()) + tuple(
+            (header or {}).items()
+        )
+        return cls(h, json.dumps(data, ensure_ascii=False).encode("utf-8"), deadline)
+
+
+@dataclass
+class Response:
+    code: int
+    header: Dict[str, str]
```

Comment:
```suggestion
    headers: Dict[str, str]
```

Reply:
A request has only one header!

What is it with the whole world 🌎first using "header fields" for "header" and then compressing the earlier to "headers" and ending up with unnecessary plural? Why is everyone crazy except me? 😭